### PR TITLE
Allow sharing assessments

### PR DESCRIPTION
**Pull request by Michael-Holmes-SDE**
note: should be merged after #10029

Allows the sharing of assessments through public course instances. TBD: Whole course instances will be shared, rather than individual assessments, or allow both?

Here's the public course instance assessments page showing all assessments for the course instance:
Screenshot 2024-08-28 120601

And the public assessment's questions page for one of the assessments:
Screenshot 2024-08-28 120842

And the public question preview page for one of the questions:
Screenshot 2024-08-28 120854

### CODE REVIEW

**Comment by Michael-Holmes-SDE** [comment]
This probably won't make sense, I'll rewrite it as I get back into the flow of this.

I'm having a difficult time figuring out how to do this. I believe it's supposed to happen when we load from disk by pressing 'Load from Disk' in the top right corner of the page:

- If a course instance is shared publicly, Iterate over all the assessments in the course instance, and iterate over all the questions in those assessments. 
   - [ X ] Make sure all those questions are shared publicly. If any of them are not, then throw an error! (check this using the sharedPublicly property of the info.json for the question)
       - This part I think is working
   - [ IP ] Save that field to the correct place in the database
       - I'm having trouble with how to save it/pass data to the sql thing (sync_course_instances.sql)
           - By default, what I'm trying to save as the 'Public' identifier of the course_instance is always null, it doesn't get the boolean from the main page

File: apps/prairielearn/src/pages/publicInstructorAssessmentQuestions/publicInstructorAssessmentQuestions.html.ts [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        ${HeadContents({ resLocals })}
+        ${compiledScriptTag('publicInstructorAssessmentQuestionsClient.ts')}
+      </head>
+      <body>
+        ${renderEjs(import.meta.url, "<%- include('../partials/navbar'); %>", resLocals)}
+        <main id="content" class="container-fluid">
+          <div class="card mb-4">
+            <div class="card-header bg-primary text-white d-flex align-items-center">
+              ${resLocals.assessment.title} ${resLocals.assessment.number}: Questions
+            </div>
```
Comment (SethPoulsen) [inline apps/prairielearn/src/pages/publicInstructorAssessmentQuestions/publicInstructorAssessmentQuestions.html.ts:70]: the whole new way of doing the Navbar was just merged, so you'll have to update this here. 

You will need to merge master and then check out this PR to see how it works on other pages: https://github.com/PrairieLearn/PrairieLearn/pull/10277

File: apps/prairielearn/src/pages/publicInstructorAssessmentQuestions/publicInstructorAssessmentQuestions.html.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+                      : html`
+                          <span class="ml-3">
+                            ${question.alternative_group_number}.${question.number_in_alternative_group}.
+                          </span>
+                        `}
+                    ${question.title}
+                    ${renderEjs(import.meta.url, "<%- include('../partials/issueBadge') %>", {
+                      urlPrefix,
+                      count: question.open_issue_count,
+                      issueQid: question.qid,
```
Comment (SethPoulsen) [inline apps/prairielearn/src/pages/publicInstructorAssessmentQuestions/publicInstructorAssessmentQuestions.html.ts:161]: I don't think we want issue counts showing up on public pages.

File: apps/prairielearn/src/pages/publicInstructorAssessments/publicInstructorAssessments.html.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  const { urlPrefix, course, __csrf_token } = resLocals;
+
+  return html`
+    <!doctype html>
+    <html lang="en">
+      <head>
+        ${HeadContents({ resLocals })} ${compiledScriptTag('instructorAssessmentsClient.ts')}
+      </head>
+      <body>
+        ${renderEjs(import.meta.url, "<%- include('../partials/navbar'); %>", resLocals)}
```
Comment (SethPoulsen) [inline apps/prairielearn/src/pages/publicInstructorAssessments/publicInstructorAssessments.html.ts:39]: same comment from above about updating the navbar

File: apps/prairielearn/src/pages/publicInstructorAssessmentQuestions/publicInstructorAssessmentQuestions.sql [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  ag.number AS alternative_group_number,
+  ag.number_choose AS alternative_group_number_choose,
+  (
+    count(*) OVER (
+      PARTITION BY
+        ag.number
+    )::integer
+  ) AS alternative_group_size,
+  z.title AS zone_title,
+  z.number AS zone_number,
```
Comment (SethPoulsen) [inline apps/prairielearn/src/pages/publicInstructorAssessmentQuestions/publicInstructorAssessmentQuestions.sql:51]: A huge amount of this query is just about getting the issue count, you should be able to get rid of all that

**Review by SethPoulsen — CHANGES_REQUESTED** [review]
This looks like a great start!

I talked with some of the other developers and everyone seems happy about the User Interface as a good starting point for these pages. 

We weren't able to come to a final decision yet about if course instances should always be shared all together, or if you should also be able to share individual assessments, but hopefully we can come to a conclusion on that in the next couple of weeks.

I left some comments on a few things that need to be fixed at this point. 

Another thing we should do is, have some automated tests for these pages. The could be pretty simple to start, basically just sync a course that has a shared course instance, then make sure you are actually able to `fetch` the URL successfully.

File: apps/prairielearn/src/pages/publicInstructorAssessmentQuestions/publicInstructorAssessmentQuestions.sql (hunk 4 above)
Comment (Michael-Holmes-SDE) [reply apps/prairielearn/src/pages/publicInstructorAssessmentQuestions/publicInstructorAssessmentQuestions.sql:51]: I removed the issue counts portion from the query and the page

File: apps/prairielearn/assets/scripts/publicInstructorAssessmentQuestionsClient.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  $('[data-toggle="popover"]').popover({ sanitize: false });
+
+  $('.js-sync-popover[data-toggle="popover"]')
+    .popover({ sanitize: false })
+    .on('show.bs.popover', function () {
+      $($(this).data('bs.popover').getTipElement()).css('max-width', '80%');
+    });
+
+  document.querySelectorAll<HTMLElement>('.js-histmini').forEach((element) => histmini(element));
+});
```
[the remark this answers is not in the document]
Comment (Michael-Holmes-SDE) [reply apps/prairielearn/assets/scripts/publicInstructorAssessmentQuestionsClient.ts:24]: I deleted the file and all references to it, and made sure everything still works

File: apps/prairielearn/src/schemas/schemas/infoAssessment.json [hunk 6]
```diff
@@ -160,6 +160,11 @@
     "module": {
       "description": "Module that this assessment belongs to, as defined in infoCourse.json.",
       "type": "string"
+    },
+    "sharedPublicly": {
```
Comment (SethPoulsen) [inline apps/prairielearn/src/schemas/schemas/infoAssessment.json:164]: we are going to use `sharedSourcePublicly` for this instead

File: database/tables/course_instances.pg [hunk 7]
```diff
@@ -8,6 +8,7 @@ columns
     id: bigint not null default nextval('course_instances_id_seq'::regclass)
     long_name: text
     ps_linked: boolean not null default true
+    shared_publicly: boolean not null default true
```
Comment (SethPoulsen) [inline database/tables/course_instances.pg:11]: do changes for course_instances in a seperate PR

File: apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        return;
+      }
+
+      console.log('serverJob', serverJob); // TEST
+
+      /*
+      const courseInstanceID = await sqldb.queryRow(
+        sql.select_course_instance_id_from_uuid,
+        { uuid: editor.uuid, course_id: res.locals.course.id },
+        IdSchema,
```
Comment (SethPoulsen) [inline apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts:125]: clean up commented code

Comment (SethPoulsen) [reply apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts:125]: and console.logs

File: apps/prairielearn/src/pages/publicQuestionPreview/publicQuestionPreview.ts [hunk 9]
```diff
@@ -37,7 +37,7 @@ async function setLocals(req, res) {
   }
 
   if (
-    !res.locals.question.shared_publicly ||
+    !res.locals.question.shared_publicly &&
```
Comment (SethPoulsen) [inline apps/prairielearn/src/pages/publicQuestionPreview/publicQuestionPreview.ts:40]: revert this change

File: apps/prairielearn/src/schemas/schemas/infoAssessment.json (hunk 6 above)
Comment (SethPoulsen) [reply apps/prairielearn/src/schemas/schemas/infoAssessment.json:164]: and shared_source_publicly in the database (similar to what is done for questions here: https://github.com/PrairieLearn/PrairieLearn/pull/10029)

File: apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+      // Get needed data
+      const courseId = await selectCourseIdByInstanceId(res.locals.course_instance_id);
+      res.locals.course = await selectCourseById(courseId);
+      res.locals.course_instance = await selectCourseInstanceById(res.locals.course_instance_id);
+      res.locals.assessment = await selectAssessmentById(res.locals.assessment_id);
+
+      res.locals.assessment.sharedPublicly = true; // TEST
+      res.locals.user = { id: 'test' }; // TEST
+      res.locals.authz_data = { authn_user: { user_id: 'test' } }; // TEST
+
```
Comment (SethPoulsen) [inline apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts:104]: clean up testing code here

File: apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.sql [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+  course_instances AS ci
+WHERE
+  ci.id = $course_instance_id;
+
+
+-- BLOCK questions
+SELECT
+  aq.*,
+  q.qid,
+  q.title,
```
Comment (SethPoulsen) [inline apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.sql:30]: This is identical to the query in `apps/prairielearn/src/pages/instructorAssessmentQuestions/instructorAssessmentQuestions.sql` except for not having the issue count. Should we instead factor this query out to `models/questions.ts` and call it `selectQuestionsForAssessment` or something like that?

File: apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        assessment_id: res.locals.assessment.id,
+        course_id: res.locals.course.id,
+      },
+      AssessmentQuestionRowSchema,
+    );
+    const questions = questionRows.map((row) => {
+      if (row.sync_errors) row.sync_errors_ansified = ansiUp.ansi_to_html(row.sync_errors);
+      if (row.sync_warnings) row.sync_warnings_ansified = ansiUp.ansi_to_html(row.sync_warnings);
+      return row;
+    });
```
Comment (SethPoulsen) [inline apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts:86]: Do we want to show these errors on the public version of the assessment or no?

I can think of arguments for both sides.

File: apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts (hunk 10 above)
Comment (Michael-Holmes-SDE) [reply apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts:104]: Do we want to remove it since we're not doing the copying functionality yet?

File: apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts (hunk 8 above)
Comment (Michael-Holmes-SDE) [reply apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts:125]: Removed

File: apps/prairielearn/src/pages/publicQuestionPreview/publicQuestionPreview.ts (hunk 9 above)
Comment (Michael-Holmes-SDE) [reply apps/prairielearn/src/pages/publicQuestionPreview/publicQuestionPreview.ts:40]: Reverted

File: apps/prairielearn/src/pages/publicQuestionPreview/publicQuestionPreview.ts [hunk 13]
```diff
@@ -52,7 +52,7 @@ router.post(
     if (req.body.__action === 'grade' || req.body.__action === 'save') {
       const variant_id = await processSubmission(req, res);
       res.redirect(
-        `${res.locals.urlPrefix}/question/${res.locals.question.id}/preview/?variant_id=${variant_id}`,
+        `${res.locals.urlPrefix}/public/course/${res.locals.course.id}/question/${res.locals.question.id}/preview/?variant_id=${variant_id}`,
```
[the remark this answers is not in the document]
Comment (Michael-Holmes-SDE) [reply apps/prairielearn/src/pages/publicQuestionPreview/publicQuestionPreview.ts:55]: This change wasn't needed and has been reverted

File: database/tables/course_instances.pg (hunk 7 above)
Comment (Michael-Holmes-SDE) [reply database/tables/course_instances.pg:11]: Removed line from this branch

File: apps/prairielearn/src/schemas/schemas/infoAssessment.json (hunk 6 above)
Comment (Michael-Holmes-SDE) [reply apps/prairielearn/src/schemas/schemas/infoAssessment.json:164]: Do we want to replace all instances of sharedPublicly  on assessments with sharedSourcePublicly?

Comment (Michael-Holmes-SDE) [reply apps/prairielearn/src/schemas/schemas/infoAssessment.json:164]: Like throughout the whole branch?

File: apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts (hunk 10 above)
Comment (Michael-Holmes-SDE) [reply apps/prairielearn/src/pages/publicAssessmentQuestionsPreview/publicAssessmentQuestionsPreview.ts]: I removed the copy button and its backend stuff

File: apps/prairielearn/src/schemas/schemas/infoAssessment.json (hunk 6 above)
Comment (Michael-Holmes-SDE) [reply apps/prairielearn/src/schemas/schemas/infoAssessment.json]: I changed the name throughout the whole branch and changed the migration accordingly

Comment (SethPoulsen) [reply apps/prairielearn/src/schemas/schemas/infoAssessment.json]: Yes, that's great thanks
